src/AutoRest.CSharp/Common/Input/Source/ProtocolCompilationInput.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using Microsoft.CodeAnalysis;
+
+namespace AutoRest.CSharp.Input.Source
+{
+    public class ProtocolCompilationInput: CompilationInput
+    {
+        private readonly List<IMethodSymbol> _methodSet = new List<IMethodSymbol>();
+
+        public ProtocolCompilationInput(Compilation compilation)
+            : base(compilation) { }//, type => type.Name.EndsWith("Client"), method => method.Parameters.Length > 0 && method.Parameters.Last().Name == "context") { }
+
+        public override void FilterSymbols()
+        {
+            foreach (IModuleSymbol module in _compilation.Assembly.Modules)
+            {
+                foreach (var type in SourceInputHelper.GetSymbols(module.GlobalNamespace))
+                {
+                    if (type is INamedTypeSymbol typeSymbol && isClient(typeSymbol))
+                    {
+                        foreach (var member in typeSymbol.GetMembers())
+                        {
+                            if (member is IMethodSymbol methodSymbol && isProtocolMethod(methodSymbol))
+                            {
+                                _methodSet.Add(methodSymbol);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        public IMethodSymbol? FindSymbol(string namespaceName, string clientName, string methodName, IEnumerable<string> parameters)
```

Comment:
First, this method is finding a method symbol, therefore I believe a better name of this should be `FindMethodSymbol`.
This method is general enough to be the real abstract method.
The idea of abstraction and polymorphism is to extract common behaviors from all possible derived types.
Finding a method symbol on an existing compilation is kind of behavior that all possible derived type of `CompilationInput` should possess, therefore on the base type, we should make this method abstract, and override it here.

Reply:
Also we should do the same on this method as the comments I put in SourceInputModel.cs
(adding descriptions, change the parameter name (clientName => className), and possibly change the type or specify the meaning of parameters)